qiskit/backends/basejob.py

```diff
@@ -0,0 +1,63 @@
+from abc import ABC, abstractmethod
+
+class BaseJob(ABC):
+    """Class to handle asynchronous jobs"""
+    
+    @abstractmethod
+    def __init__(self):
+        """Initializes and initates the asynchronous job"""
+        pass
+
+    @abstractmethod
+    def result(self):
+        """
+        Returns:
+            qiskit.Result:
+        """
+        pass
+
+    @abstractmethod
+    def status(self):
+        """
+        Returns:
+            str: 'ERROR', 'QUEUED', 'RUNNING', 'CANCELLED', or 'DONE'
```

Comment:
While we are at it, can we introduce something a bit more robust than plain strings, such as `Enum`?

Reply:
Actually the spec describes status as a dictionary which I'm changing this to. The status dictionary looks like,
```
    @property
    @abstractmethod
    def status(self):
        """
        Returns:
            dict: {'job_id': <job_id>,
                   'status': 'ERROR'|'QUEUED'|'RUNNING'|'CANCELLED'|'DONE',
                   'status_msg': <str>
        """
        pass
```
I suppose that `status` key could be `Enum` however that could be a hassle if we need to serialize it at some point.

Reply:
`Enums` have some advantages over strings in this kind of contexts:
* It's safer and less prone to user typing errors.
* IDEs usually have a proper support.
* If there's a need to change the value of the `Enum`, you just need to change it in one place.
* Less memory usage (not a big deal for this class, though)

I wouldn't worry to much about the serialization, depending on the type of serialization, the `Enum` will turn into a string or a binary value.

Reply:
`status` is an `enum` in the schema (as of now)

Reply:
@atilag the thing about the `job_id` is that the physical thing that runs and the result it produces are not part of `qiskit`. Technically some one could use a different program to run a job and how would we be able to grab the results (or vice-versa)?

Reply:
Any remote job will need a job_id, maybe not local simulators (but I could still think of cases, e.g., local simulators with threading that could make use of a job_id)

Reply:
If you want to leave it out of the qiskit architecture that could be ok, but it will need to be in the result data structure that is returned by the job